File: app/src/main/java/com/govnokoder/velotracker/TrainingActivity.java
package com.govnokoder.velotracker;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.location.Location;
import android.os.Bundle;
import android.os.IBinder;
import android.view.View;
import android.view.WindowManager;
import android.widget.ProgressBar;

import com.google.android.material.tabs.TabLayout;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.preference.PreferenceManager;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayoutMediator;
import com.govnokoder.velotracker.BL.ParcelableTraining;
import com.govnokoder.velotracker.messages.SharedViewModel;
import com.govnokoder.velotracker.services.LocationService;
import com.govnokoder.velotracker.ui.training.PageMap;
import com.govnokoder.velotracker.ui.training.ViewPagerAdapterTr;

public class TrainingActivity extends AppCompatActivity implements PageMap.onSomeEventListener {

    TabLayout tabLayout;
    ViewPager2 viewPager;
    ViewPagerAdapterTr viewPagerAdapterTr;
    ProgressBar progressBar;

    private MyReceiver myReceiver;

    private LocationService mService = null;
    private boolean isStart = false;

    private boolean mBound = false;

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        startActivity(new Intent(getApplicationContext(), MainActivity.class));
        finish();
    }

    private final ServiceConnection mServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            LocationService.LocalBinder binder = (LocationService.LocalBinder) service;
            mService = binder.getService();
            mBound = true;
        }
        @Override
        public void onServiceDisconnected(ComponentName name) {
            mService = null;
            mBound = false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_training);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        viewPager = findViewById(R.id.view_pager);
        viewPagerAdapterTr = new ViewPagerAdapterTr(this);
        viewPager.setAdapter(viewPagerAdapterTr);

        tabLayout = findViewById(R.id.tabs);
        TabLayoutMediator tabLayoutMediator = new TabLayoutMediator(tabLayout, viewPager, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                switch (position) {
                    case 0:
                        tab.setText("1");
                        break;
                    case 1:
                        tab.setText("2");
                }
            }
        });
        tabLayoutMediator.attach();
        myReceiver = new MyReceiver();
        startForegroundService(new Intent(this, LocationService.class));
        progressBar = findViewById(R.id.progressBar);
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    protected void onStart() {
        super.onStart();
        bindService(new Intent(this, LocationService.class), mServiceConnection,
                Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        LocalBroadcastManager.getInstance(this).registerReceiver(myReceiver,
                new IntentFilter(LocationService.ACTION_BROADCAST));
    }

    @Override
    protected void onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(myReceiver);
        super.onPause();
    }

    @Override
    protected void onStop() {
        if (mBound) {
            unbindService(mServiceConnection);
            mBound = false;
        }
        super.onStop();
    }

    private void sendParcelableTraining(ParcelableTraining parcelableTraining){
        SharedViewModel model = new ViewModelProvider(this).get(SharedViewModel.class);
        model.sendMessage(parcelableTraining);
    }

    @Override
    public void onPauseTraining() {
        mService.onPause();
    }

    @Override
    public void onStopTraining(boolean isSave) {
        mService.onStop(isSave);
    }

    @Override
    public void onResumeTraining() {
        mService.onResume();
    }

    public class MyReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            ParcelableTraining parcelableTraining = intent.getParcelableExtra(LocationService.EXTRA_PARCELABLE_TRAINING);
            if (parcelableTraining != null) {
                if(!isStart){
                    progressBar.setVisibility(View.INVISIBLE);
                }
                sendParcelableTraining(parcelableTraining);
            }
        }
    }


}

File: app/src/main/java/com/coursework/velotracker/TrainingActivity.kt
package com.coursework.velotracker

import android.content.*
import android.os.Bundle
import android.os.IBinder
import android.view.View
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.coursework.velotracker.BL.Model.Training.ParcelableTraining
import com.coursework.velotracker.ViewModels.SharedViewModel
import com.coursework.velotracker.Services.LocationService
import com.coursework.velotracker.databinding.ActivityTrainingBinding
import com.coursework.velotracker.ui.fragments.training.PageMap
import com.coursework.velotracker.ui.fragments.training.ViewPagerAdapter
import com.google.android.material.tabs.TabLayoutMediator

class TrainingActivity: AppCompatActivity(), PageMap.OnSomeEventListener {

    private lateinit var binding: ActivityTrainingBinding

    lateinit var viewPagerAdapter: ViewPagerAdapter
    private lateinit var myReceiver:MyReceiver
    private var mService: LocationService? = null
    private val isStart = false
    private var mBound = false

    override fun onBackPressed() {
        super.onBackPressed()
        startActivity(Intent(applicationContext, MainActivity::class.java))
        finish()
    }

    private val mServiceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val binder = service as LocationService.LocalBinder
            mService = binder.getService()
            mBound = true
        }
        override fun onServiceDisconnected(name: ComponentName) {
            mService = null
            mBound = false
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityTrainingBinding.inflate(layoutInflater)
        setContentView(binding.root)
    }

    override fun onStart() {
        super.onStart()
        initFunc()
        bindService(Intent(this, LocationService::class.java), mServiceConnection, Context.BIND_AUTO_CREATE)
    }

    private fun initFunc(){
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        viewPagerAdapter = ViewPagerAdapter(this)
        binding.viewPager.adapter = viewPagerAdapter
        val tabLayoutMediator = TabLayoutMediator(binding.tabs, binding.viewPager) { tab, position ->
            when (position) {
                0 -> tab.text = "1"
                1 -> tab.text = "2"
            }
        }
        tabLayoutMediator.attach()
        myReceiver = MyReceiver()
        startForegroundService(Intent(this, LocationService::class.java))
        binding.progressBar.visibility = View.VISIBLE
    }

    override fun onResume() {
        super.onResume()
        LocalBroadcastManager.getInstance(this).registerReceiver(myReceiver, IntentFilter(LocationService.ACTION_BROADCAST))
    }

    override fun onPause() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(myReceiver)
        super.onPause()
    }

    override fun onStop() {
        if (mBound) {
            unbindService(mServiceConnection)
            mBound = false
        }
        super.onStop()
    }

    private fun sendParcelableTraining(parcelableTraining: ParcelableTraining) {
        val model: SharedViewModel = ViewModelProvider(this).get<SharedViewModel>(SharedViewModel::class.java)
        model.sendMessage(parcelableTraining)
    }

    override fun onPauseTraining() {
        mService!!.onPause()
    }

    override fun onStopTraining(isSave: Boolean) {
        mService!!.onStop(isSave)
    }

    override fun onResumeTraining() {
        mService!!.onResume()
    }

    inner class MyReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent) {
            val parcelableTraining: ParcelableTraining? = intent.getParcelableExtra(LocationService.EXTRA_PARCELABLE_TRAINING)
            if (parcelableTraining != null) {
                if (!isStart) {
                    binding.progressBar.visibility = View.INVISIBLE
                }
                sendParcelableTraining(parcelableTraining)
            }
        }
    }
}